StandardLibrary/src/Sequence.dfy

```diff
@@ -0,0 +1,41 @@
+// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+include "UInt.dfy"
+
+module StandardLibrary.Sequence {
+  // export provides SequenceEqual, StandardLibrary
+  import opened UInt
+
+  predicate method SequenceEqualNat<T(==)>(seq1: seq<T>, seq2: seq<T>, start1: nat, start2: nat, size: nat) : (ret : bool)
+    requires start1 + size <= |seq1|
+    requires start2 + size <= |seq2|
+    ensures ret ==> seq1[start1..start1 + size] == seq2[start2..start2 + size]
+  {
+    if |seq1| > INT64_MAX_LIMIT || |seq2| > INT64_MAX_LIMIT then
+      false
```

Comment:
Issue: False, here on line 16, signals the Sequences are not equal.

But actually, we do not know that.
All we know is the sequences are very large.

Should we change `SequenceEqualNat` to be `function method`,
such that it return a `Result<bool, Error>` and error on the sequences greater than INT64_MAX,
throwing/returning an Unsupported Operation Exception
(or some semantic equivalent of that)?
 
Or, should we move the requires from lines 22 & 23 up to here,
and say that until we/Dafny has non-INT_64 max support,
we require this?

And then the callers can decide if they want to throw an exception,
or prove that their sequences are of a reasonable size.

My preference is for moving the requires up.
I think most callers will be able to prove their sequences are smaller than 
`9,223,372,036,854,775,807`. 

For example,
I think that exceeds the number characters of any stack trace I have ever read.

The only time I think I have ever had to work with so many characters is
doing log diving across multiple services at once.

At which point,
you use dedicated to tools to execute the string search.

Reply:
It's unsigned, so the number is actually `18,446,744,073,709,551,616`. That's more that will fit in the RAM of any computer anywhere, and would take longer than the age of the universe to check for the substring. I guarantee that your log diving did not involve 18 exabytes of data. The Dafny team has even agreed with me, in principle, to one day generate native code with this very assumption -- that no actual sequence in a real running program will ever exceed 18,446,744,073,709,551,616 entries.

The problem with Dafny (one of the many) is that proving your sequence is smaller than 18,446,744,073,709,551,616 entries is no easier than proving that it's smaller than 42 entries.

It is perfectly valid to assume that no sequence will ever exceed 18,446,744,073,709,551,616 entries.

Reply:
If they can prove it, they can use `SequenceEqual` directly.
If they can't prove it, then a runtime check is needed, and better in one place than in many places.

Reply:
I stand corrected.
Sorry, I had a poor understanding of the magnitude of these numbers,
and I just saw a logical/CX issue.

i.e: `false` instead of `Not Supported`. 

An alternative suggestion:
```
datatype SequenceEquality = 
    | TrueValue
    | FalseValue
    | UnSupported
 ```

This makes everything very clear.
But I am not blocking on this,
and will happily approve.